async_server.c

```diff
@@ -0,0 +1,364 @@
+#include "async_server.h"
+
+#include "brute.h"
+#include "log.h"
+#include "multi.h"
+#include "server_common.h"
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <pthread.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+static acl_context_t *
+acl_context_init (serv_context_t *global_ctx)
+{
+  acl_context_t *ctx = calloc (1, sizeof (*ctx));
+  if (!ctx)
+    {
+      error ("Could not allocate memory for client context");
+      return (NULL);
+    }
+  ctx->context = global_ctx;
+
+  if (queue_init (&ctx->registry_idx, sizeof (int)) != QS_SUCCESS)
+    {
+      error ("Could not initialize registry indices queue");
+      return (NULL);
+    }
+  for (int i = 0; i < QUEUE_SIZE; ++i)
+    if (queue_push (&ctx->registry_idx, &i) != QS_SUCCESS)
+      {
+        error ("Could not push index to registry indices queue");
+        return (NULL);
+      }
+
+  ctx->ref_count = 2;
+  if (pthread_mutex_init (&ctx->mutex, NULL) != 0)
+    {
+      error ("Could not initialize mutex");
+      return (NULL);
+    }
+
+  return ctx;
+}
+
+static void
+acl_context_destroy (acl_context_t *ctx)
+{
+  trace ("Destroying client context");
+
+  if (queue_cancel (&ctx->registry_idx) != QS_SUCCESS)
+    {
+      error ("Could not cancel registry indices queue");
+      goto cleanup;
```

Comment:
we still need to free up queue memory... let's just report errors and continue with normal workflow

Reply:
Done